[FEAT] Add rewards API

benjaminkostiuk opened the pull request:
## Overview

<!-- Give a brief overview of your changes. -->

Implements the endpoints for rewards api.

Still missing tests to be implemented in next PR.

This change is a

- [ ] Bug fix
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that will change exisiting functionality)
- [ ] Documentation update

## Description

<!-- Describe your changes in detail -->
<!-- Enumerate all
 - Areas affected
 - Features added
 - Tests added -->
 - New service methods and methods and errors for rewards.

## Motivation/Links
One step closer to finishing the rewards api.
<!-- Why was this change needed? -->
<!-- Add any relevant links here, issues, cards or other pull requests. -->

ruler225 commented on app/common.py:
```diff
@@ -7,6 +7,7 @@
+from typing import Dict, Any
 
 
 class FirebaseDataEntity(ABC):
```
I'm cool with ditching this class, but the to_json function needs to be added back to whatever data classes were inheriting from it before since this was being used.

ruler225 replied:
On second thought, this is basically only used in the test so it's not necessarily needed

benjaminkostiuk replied:
If it's okay I'll leave the removal to you so you can de-couple your stuff and I don't want to add more to the scope of this PR.

ruler225 commented on app/rewards/api.py:
```diff
@@ -0,0 +1,85 @@
+        return ReferralCodeNotFound(err_msg).jsonify()
+
+    if user_with_code.email == user.email:
+        err_msg = "Invalid operation. Refer refer yourself."
```
Cannot* refer yourself, minor typo

benjaminkostiuk replied:
Updated, nice catch!

ruler225 commented on app/rewards/api.py:
```diff
@@ -0,0 +1,85 @@
+
+
+@with_auth_user
+def submit_referral_code(user: User, code: str, **kwargs):
```
Add a flag to indicate that a user has already used a referral code to prevent abuse

benjaminkostiuk replied:
Added a new flag on the user `hasUsedReferralCode` that will be set to true once they use their code.

ruler225 commented on app/rewards/errors.py:
```diff
@@ -0,0 +1,21 @@
+        self.message = message
+
+    def jsonify(self) -> Tuple[Dict[str, str], int]:
+        return {"error": "Referral code not found", "message": self.message}, 404
```
Use predefined build_error() function

benjaminkostiuk replied:
Updated!

ruler225 commented on app/rewards/spec.yaml:
```diff
@@ -0,0 +1,117 @@
+            application/json:
+              schema:
+                $ref: "#/components/schemas/ErrorResponse"
+      security:
```
If we already know we need to secure all the endpoints in rewards, this can also be added globally to the whole file instead of each individual endpoint

benjaminkostiuk replied:
Updated to add security to all paths except health check.

ruler225 reviewed (commented):
Just have a couple comments, but nothing too severe to block merging. Just don't forget to address the referral code redemption abuse issue

benjaminkostiuk commented:
@ruler225 I've addressed comments. I'm gonna leave the following items for later PRs since this one is so big:
- Fixing the tests so they run on CI
- Adding tests for these service methods and endpoints
- Adding a sign up endpoint.

ruler225 reviewed (approved):
Nice, looks good
